src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs

```diff
@@ -40,12 +41,34 @@ internal sealed partial class ImmutabilityContext {
 			SpecialType.System_ValueType
 		);
 
-		private ImmutabilityContext( IEnumerable<ImmutableTypeInfo> extraImmutableTypes ) {
-			m_extraImmutableTypes = extraImmutableTypes
-				.ToImmutableDictionary(
-					info => info.Type,
-					info => info
-				);
+		private ImmutabilityContext(
+			ImmutableDictionary<INamedTypeSymbol, ImmutableTypeInfo> extraImmutableTypes,
+			ImmutableHashSet<ITypeParameterSymbol> conditionalTypeParamemters
+		) {
+			m_extraImmutableTypes = extraImmutableTypes;
+			m_conditionalTypeParameters = conditionalTypeParamemters;
+		}
+
+		/// <summary>
+		/// This function is intended to provide a new immutability context that considers the conditional type parameters
+		/// of the ConditionallyImmutable type being checked as immutable. It is important this is only used in those cases,
+		/// and not for instance while validating type arguments to an [Immutable] type parameter.
+		/// </summary>
+		/// <param name="type">The ConditionallyImmutable type definition being checked</param>
+		/// <returns></returns>
+		public ImmutabilityContext WithConditionalTypeParametersAsImmutable(
+			INamedTypeSymbol type
+		) {
+			if( !Attributes.Objects.ConditionallyImmutable.IsDefined( type ) ) {
+				throw new InvalidOperationException( $"{nameof( WithConditionalTypeParametersAsImmutable )} should only be called on ConditionallyImmutable types" );
+			}
+
+			var conditionalTypeParameters = type.TypeParameters.Where( p => Attributes.Objects.OnlyIf.IsDefined( p ) );
+
+			return new ImmutabilityContext(
+				extraImmutableTypes: m_extraImmutableTypes,
+				conditionalTypeParamemters: m_conditionalTypeParameters.Union( conditionalTypeParameters )

```

Comment:
Union is never actually needed right now, but 🤷

Reply:
Nested classes?

Reply:
```cs
[ConditionallyImmutable]
class Parent<[OnlyIf]T> {
  class Child<[OnlyIf]U> {
    private readonly T m_t;
    private readonly U m_u;
    // ...
  }
}
```

Reply:
Yeah, as in, not trying to handle nested classes.